### DISCUSSION TITLE
Client components having server rendered children?

### DISCUSSION THREAD

**Discussion by onursagir** [vercel/next.js#43153] [Help]
This figure ![image](https://user-images.githubusercontent.com/2290276/202918095-fb78e706-d815-45e2-8b1c-628a7ebafa01.png) on   [Component-level Client and Server Rendering](https://beta.nextjs.org/docs/rendering/fundamentals#component-level-client-and-server-rendering) makes it seem that client component can have children that are server components is this really the case? And if so how does NextJs go about figuring out what the children are of a client component without rendering it first.

**Comment by icyJoseph** [comment]
It is done through the so called "composition" technique. A Client component can have a "slot" for children. Later on, when you create the client component element via `<ClientComponent></ClientComponent>` JSX, you can slot in a Server Component as children:

```jsx
import ClientComponent from "./ClientComponent";
import ServerComponent from "./ServerComponent";

// Pages are Server Components by default
export default function Page() {
  return (
    <ClientComponent>
      <ServerComponent />
    </ClientComponent>
  );
}
```

Documentation can be found here: https://beta.nextjs.org/docs/rendering/server-and-client-components#importing-server-components-into-client-components.

And here's a deep dive into composition, https://www.youtube.com/watch?v=3XaXKiXtNjw 

Note that if the server component is just something like, `const Hi = () => <div>hi</div>`, you might be able to directly use it as children to a client component. As soon as you try to use more complex stuff, the compiler will see that its parent is a client component and some guards kick in.

**Reply by onursagir** [reply]
After the release of the stable version of app router, I was wondering if this is still the case mainly asking because of this https://nextjs.org/docs/getting-started/react-essentials#nesting-server-components-inside-client-components

**Reply by icyJoseph** [reply]
What about it? It hasn't changed. You still need to compose them with the children prop. 

One clarification is that if you nest a component that doesn't have `use client`, into a client component, directly, it'll be opted in to client component architecture. Most notably doing this with async components can lead to various problems. 

That's why it is called an unsupported pattern. 

So you have to just go with the children props/slot approach, aka composition.

**Reply by icyJoseph** [reply]
Though I must also add this https://www.youtube.com/watch?v=ciQdQkT0bZk https://github.com/dai-shi/waku, which can do what you wanted.

**Comment by kotAPI** [comment]
There are also instances where you'd want to force client-side only behaviour for some components, for instance, a third party UI library that I use doesn't play well with next13, i'd have my wrapper declared as client, if all the children render as client too - then I don't have to worry about it being rendered on the server, which is the case right now.. Is there anyway I can render my 3rd party UI library only on the client?

I'm specifically talking about antd library here, but it may as well apply to any other library that is not on react 18. Any workarounds for this? I'm having a tough time migrating the components.

**Reply by aidanCQ** [reply]
You may need to rexport your third party library and mark the export file with 'use client'. I had issues with MUI in the past although I believe they've marked all their components with 'use client'.

**Reply by kotAPI** [reply]
But won't that component be specifically rendered as an RSC?

**Reply by kotAPI** [reply]
If you look at OPs post of components hierarchy, the component that's wrapped will act as a client, but the one we export will still be an RSC right? Correct me if I'm wrong

**Reply by aidanCQ** [reply]
I'm not sure I follow, sorry. If you want your third party library to render on the client, you'll need to introduce a 'use client' directive somewhere. I don't think there is a way to force all children to render as client components because there could be server rendered children. You may need to declare 'use client' in each child page. If you re-export your client library as client components you can still use server rendering and only client render those components. Hope this helps. Please ask others as well as my knowledge is a little rusty.

**Reply by kotAPI** [reply]
yeah I meant the same thing, we're both on the same page. 

I have a bit of trouble understand how the RSCs work in this context as I tried the same thing you suggested and couldn't get it to work.
